src/backend/app/tools.py

```diff
@@ -61,19 +62,79 @@ def generate_prediction(eval_params):
         {h3.geo_to_h3(lat, lon, hex_resolution) for lat, lon in coordinates}
     )
 
-    hull = alphashape.alphashape(coordinates, 1)
-    hull_points = list(mapping(hull)["coordinates"])
+    # hull = alphashape.alphashape(coordinates, 1)
+    # hull_points = list(mapping(hull)["coordinates"])
 
     annotation_hexagon_ids = {
         "presence": prediction_hexagon_ids,
         "absence": list()
     }
 
     return dict(
-        coordinates=coordinates.tolist(),
-        pred_loc_combined=pred_loc_combined.tolist(),
-        hull_points=hull_points,
+        # coordinates=coordinates.tolist(),
+        # pred_loc_combined=pred_loc_combined.tolist(),
+        # hull_points=hull_points,
         prediction_hexagon_ids=prediction_hexagon_ids,
         # set prediction_hexagon_ids as the starting point for the annotation
         annotation_hexagon_ids=annotation_hexagon_ids,
     )
+
+
+def generate_prediction_scores(eval_params, lowest_threshold=0.0001):
+    preds, locs = get_prediction(eval_params)
+
+    # switch lats and longs
+    locs[:, [1, 0]] = locs[:, [0, 1]]
+
+    # combine coordinates and predictions
+    pred_loc_combined = np.column_stack((locs, preds))
+    pred_loc_combined = np.float_(pred_loc_combined)
+    pred_loc_combined = pred_loc_combined[pred_loc_combined[:, 2] >= lowest_threshold]
+    hex_resolution = eval_params['hex_resolution']
+
+    # if a more detailed HeatMap needed, use `pred_loc_combined` for that
+    scores_dict={}
+    for lat, lon, pred in pred_loc_combined:
+        try:
+            scores_dict[h3.geo_to_h3(lat, lon, hex_resolution)].append(pred)
+        except Exception:
+            scores_dict[h3.geo_to_h3(lat, lon, hex_resolution)]=[pred]
+    hex_indexes=[]
+    hex_scores=[]
+    for key in scores_dict:
+        hex_indexes.append(key)
+        hex_scores.append(max(scores_dict[key]))
+    
+    index_score_combined = np.column_stack((hex_indexes, hex_scores))
+
+    return index_score_combined
+
+
+def populate_prediction_database(eval_params,db):
+    index_score_combined = generate_prediction_scores(eval_params)
+
+    prediction_model = models.Prediction()
+    prediction_model.taxa_id = eval_params["taxa_id"]
+    db.add(prediction_model)
+    db.commit()
+
+    for index_score in index_score_combined:
+        prediction_hexagon_model = models.PredictionHexagon()
+        prediction_hexagon_model.prediction_id = prediction_model.prediction_id
+        prediction_hexagon_model.hex_index = index_score[0]
+        prediction_hexagon_model.hex_score = index_score[1]
+        db.add(prediction_hexagon_model)
+    db.commit()    
+
+    return index_score_combined
+
+def get_predicted_hexagons(db, eval_params):
+    taxa_name = eval_params["taxa_name"]
+    taxa_id = int(taxa_name.split("(")[-1][:-1])
+    try:
```

Comment:
I would recommend using `try ... except` only if you really need it.
Now all errors are caught here, so we may not notice an error that is not related to the lack of prediction in the database

It's easier to check if prediction was found in the database, and than find hexagons, otherwise, return None

here is how I searching for the latest annotation  in db:
```
    latest_annotation = (
        db.query(models.Annotation)
        .filter(models.Annotation.taxa_id == taxa_id)
        .order_by(models.Annotation.created_at.desc())
        .first()
    )
```
you can do something like this, it looks more clear for me...

Reply:
This is a good point. It's good to make exceptions specific, so if there's a particular type of exception you know you want to handle here, just handle that and let all other types of exceptions get raised so that they don't fail silently.

Reply:
@SergePogorelov This was what I was trying to do but wasn't sure, used your solution and it's working properly. Thank you so much!
@ginic Noted! Will get rid of the other try except I have.